src/Data/List/Properties.agda

```diff
@@ -746,17 +746,41 @@ map-∷= (x ∷ xs) zero    v f = refl
 map-∷= (x ∷ xs) (suc k) v f = cong (f x ∷_) (map-∷= xs k v f)
 
 ------------------------------------------------------------------------
--- _─_
+-- insert
 
-length-─ : ∀ (xs : List A) k → length (xs ─ k) ≡ pred (length xs)
-length-─ (x ∷ xs) zero        = refl
-length-─ (x ∷ y ∷ xs) (suc k) = cong suc (length-─ (y ∷ xs) k)
+length-insert : ∀ (xs : List A) (i : Fin (suc (length xs))) v → length (insert xs i v) ≡ suc (length xs)
+length-insert []       zero    v = refl
+length-insert (x ∷ xs) zero    v = refl
+length-insert (x ∷ xs) (suc i) v = cong suc (length-insert xs i v)
 
-map-─ : ∀ xs k (f : A → B) →
-        let eq = sym (length-map f xs) in
-        map f (xs ─ k) ≡ map f xs ─ cast eq k
-map-─ (x ∷ xs) zero    f = refl
-map-─ (x ∷ xs) (suc k) f = cong (f x ∷_) (map-─ xs k f)
+------------------------------------------------------------------------
+-- remove
+
+length-remove : ∀ (xs : List A) k → suc (length (remove xs k)) ≡ length xs
+length-remove (x ∷ xs) zero        = refl
+length-remove (x ∷ y ∷ xs) (suc k) = cong suc (length-remove (y ∷ xs) k)
+
+map-remove : ∀ xs k (f : A → B) →
+            let eq = sym (length-map f xs) in
+            map f (remove xs k) ≡ remove (map f xs) (cast eq k)
+map-remove (x ∷ xs) zero    f = refl
+map-remove (x ∷ xs) (suc k) f = cong (f x ∷_) (map-remove xs k f)
+
+------------------------------------------------------------------------
+ -- insert and remove
+
+remove-insert : ∀ (xs : List A) (i : Fin (suc (length xs))) v → remove (insert xs i v) ((cast (sym (length-insert xs i v)) i)) ≡ xs
+remove-insert []       zero    v = refl
+remove-insert (x ∷ xs) zero    v = refl
+remove-insert (x ∷ xs) (suc i) v = cong (_ ∷_) (remove-insert xs i v)
+
+insert-remove : (xs : List A) (i : Fin (length xs)) → insert (remove xs i) (cast (sym (length-remove xs i)) i) (lookup xs i) ≡ xs
+insert-remove (x ∷ xs) zero = h xs x
+  where
+  h : ∀ (xs : List A) v → insert xs zero v ≡ v ∷ xs
+  h []       v = refl
+  h (x ∷ xs) v = refl
```

Comment:
I don't know why agda does not like -
```agda
insert-remove (x ∷ xs) zero = refl
```

Reply:
Hmm I assume somehow `cast` is blocking the reduction behaviour. @jamesmckinna this might be the sort of thing you're interested in working out?

Reply:
I will certainly take a look, but I think similar things came up for @shhyou in #2045 [see discussion here](https://github.com/agda/agda-stdlib/pull/2045#discussion_r1292577189) so you might like to look at that discussion first. In theory, `cast` is great, because it avoids having to match on `refl` in proofs (`--without-K` hovering off-stage, plus the desire for 'only computational proof normalisation on *data*') but in practice, once it appears in a non-trivial goal involving a list `xs` and a function (such as: `length`!) defined on the *same* list, then such coupling gives rise to 'wicked' complexity. 

Possible 'uncoupling' solutions (might) include:
* introducing additional equalities on lists, to standardise apart otherwise coupled occurrences;
* doing `with ... in ... eq` abstraction over the list `length`, in order to obtain a fresh identifier for that length on which you can do pattern-matching/induction
* reconsidering the definition of `insert`, in this instance (!): I agree that you want the lemma witnessed by `h` to be a *definitional* property...

That's my 2centimes' worth of thought (without really reading the code in detail ... yet) to be going on with, but I will look at it more closely now... thanks @MatthewDaggitt for the nudge.

Reply:
What happens if you telescope the first two lines of the definition of `insertAt` into the following?
```agda
insertAt xs zero v = v ∷ xs
```
Type/termination-checker both seem happy? And then your `h` would, indeed, be definitional!?

Reply:
Thanks for looking at this @JacquesCarette and @MatthewDaggitt! 

The following -
```agda
insertAt xs zero v = v ∷ xs
```
does solve a lot of problems!